node/src/components/chain_synchronizer/operations.rs

```diff
@@ -1289,11 +1295,199 @@ async fn fetch_block_worker(
+            Err(_) => return maybe_signatures,
+        }
+    }
+
+    Err(FetcherError::TimedOut {
+        id: block_hash,
+        peer,
+    })
+}
+
+async fn fetch_finality_signatures_by_block_header(
+    block_header: BlockHeader,
+    ctx: &ChainSyncContext<'_>,
+) -> Result<Option<BlockSignatures>, Error> {
+    let start = Timestamp::now();
+    let mut peer_list = ctx
+        .effect_builder
+        .get_fully_connected_non_joiner_peers()
+        .await;
+    ctx.filter_bad_peers(&mut peer_list);
+
+    let mut sig_collector = BlockSignaturesCollector::new();
+
+    let block_header_hash = block_header.hash(ctx.config.verifiable_chunked_hash_activation());
+
+    for peer in peer_list {
+        let fetched_signatures = fetch_finality_signatures_with_retry(
+            block_header_hash,
+            peer,
+            FINALITY_SIGNATURE_FETCH_RETRY_COUNT,
+            ctx,
+        )
+        .await;
+        let maybe_signatures = match fetched_signatures {
+            Ok(FetchedData::FromStorage { item, .. }) => {
+                trace!(
+                    ?block_header_hash,
+                    ?peer,
+                    "did not get FinalitySignatures from peer, got from storage instead",
+                );
+                Some(*item)
+            }
+            Ok(FetchedData::FromPeer { item, .. }) => {
+                trace!(?block_header_hash, ?peer, "fetched FinalitySignatures");
+                Some(*item)
+            }
+            Err(_) => None,
+        };
+
+        sig_collector.add_signatures(maybe_signatures);
+
+        let era_for_validators_retrieval =
+            get_era_id_for_validators_retrieval(&block_header.era_id());
+
+        let maybe_switch_block_of_previous_era = ctx
+            .effect_builder
+            .get_switch_block_header_at_era_id_from_storage(era_for_validators_retrieval)
+            .await;
+
+        match maybe_switch_block_of_previous_era {
+            Some(switch_block_of_previous_era) => {
+                match switch_block_of_previous_era.next_era_validator_weights() {
+                    Some(validator_weights) => {
+                        if sig_collector.check_if_sufficient(
+                            validator_weights,
+                            ctx.config.finality_threshold_fraction(),
+                        ) {
+                            info!(
+                                ?block_header_hash,
+                                height = block_header.height(),
+                                ?era_for_validators_retrieval,
+                                "fetched sufficient finalty signatures"
+                            );
+                            break;
+                        }
+                    }
+                    None => {
+                        return Err(Error::HitGenesisBlockTryingToGetTrustedEraValidators {
+                            trusted_header: block_header,
+                        });
+                    }
```

Comment:
I seem to remember a comment suggesting `.ok_or(Error::...)?` instead of these `None` branches, but it's no longer here - was that not feasible for some reason, or did the comment disappear because the code changed so much since then? If the latter, I think using `ok_or` might be nicer.

Reply:
You remember correctly, there was a comment, but the code got removed. This one is handled in 6541b1f5ab3283a44cbf27963bc78e3cf04bb8e3